### PR TITLE
Add menu configuration for entities

### PR DESCRIPTION
**Pull request by mshima**
Implement menu configuration to the entity.

This will allow further entity customization #10160.

-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: test/blueprint/client-menu-blueprint.spec.js [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+            });
+
+            /*
+            it('Force print file', () => {
+                assert.fileContent(`${CLIENT_MAIN_SRC_DIR}i18n/en/global.json`, /fail to debug/);
+            });
+            it('Force print file', () => {
+                assert.fileContent(`${CLIENT_MAIN_SRC_DIR}app/layouts/navbar/navbar.component.html`, /fail to debug/);
+            });
+            */
```
Comment (MathieuAA) [inline test/blueprint/client-menu-blueprint.spec.js]: if these commented tests aren't relevant, can you remove them?

File: test/blueprint/client-menu-blueprint.spec.js [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+
+            it('Menu changed to the admin by the blueprint', () => {
```
Comment (MathieuAA) [inline test/blueprint/client-menu-blueprint.spec.js]: can you use the same test-coding style as before? `it('changes...')`

File: test/blueprint/client-menu-blueprint.spec.js [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+                assert.noFileContent(`${CLIENT_MAIN_SRC_DIR}app/layouts/navbar/navbar.component.html`, /global.menu.entities.foo/);
+                assert.fileContent(`${CLIENT_MAIN_SRC_DIR}app/layouts/navbar/navbar.component.html`, /global.menu.admin.foo/);
+            });
+
+            it('Menu translation changed to the admin by the blueprint', () => {
```
Comment (MathieuAA) [inline test/blueprint/client-menu-blueprint.spec.js]: can you use the same test-coding style as before? `it('changes...')`

File: test/blueprint/client-menu-blueprint.spec.js [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+
+            it('Contains the menu', () => {
+                assert.fileContent(`${CLIENT_MAIN_SRC_DIR}app/layouts/navbar/navbar.component.html`, /global.menu.entities.foo/);
+            });
+
+            it('Contains the menu translation', () => {
+                assert.JSONFileContent(`${CLIENT_MAIN_SRC_DIR}i18n/en/global.json`, {
+                    global: { menu: { entities: { foo: 'Foo' } } }
+                });
+            });
```
Comment (MathieuAA) [inline test/blueprint/client-menu-blueprint.spec.js]: same comments as before:
  - commented tests
  - `it('changes...')`

File: generators/generator-base.js [hunk 5]
```diff
@@ -199,6 +304,23 @@ module.exports = class extends PrivateBase {
...
+     * @param {string} value - Default translated value
+     * @param {string} language - The language to which this translation should be added
+     */
+    addMenuTranslationKey(key, value, language, entityModule) {
+        if (entityModule === undefined || entityModule === 'entities') {
+            this.addEntityTranslationKey(key, value, language);
+        } else if (entityModule === 'admin') {
+            this.addAdminElementTranslationKey(key, value, language);
+        } else {
+            this.log(`Entry ${key} not added to translation on module ${entityModule}`);
```
Comment (MathieuAA) [inline generators/generator-base.js]: maybe adding a little bit more details here about why the key cannot be added? like 'the entity modyle isn't recognized' or something

File: generators/generator-base.js [hunk 6]
```diff
@@ -199,6 +304,23 @@ module.exports = class extends PrivateBase {
         this.needleApi.clientI18n.addEntityTranslationKey(key, value, language);
     }
 
+    /**
+     * Add a new menu entry in the "global.json" translations.
+     *
+     * @param {string} key - Key for the entity name
+     * @param {string} value - Default translated value
+     * @param {string} language - The language to which this translation should be added
+     */
```
Comment (MathieuAA) [inline generators/generator-base.js]: there's a missing parameter: entityModule

File: generators/generator-base.js [hunk 7]
```diff
@@ -199,6 +304,23 @@ module.exports = class extends PrivateBase {
...
+    addMenuTranslationKey(key, value, language, entityModule) {
```
Comment (MathieuAA) [inline generators/generator-base.js]: Refactor hint:
  - there are 4 parameters (which may be too much)
  - everytime we use the entityModule param to determine what method to call with the other 3
  - the three params might belong to the same object (or structure) and passing them in that order, everytime, isn't that pretty

File: generators/generator-base.js [hunk 8]
```diff
@@ -101,15 +101,33 @@ module.exports = class extends PrivateBase {
...
 
     /**
-     * Add a new entity in the TS modules file.
+     * Add a new entity in the "entities" menu.
+     *
+     * @param {string} routerName - The name of the Angular router (which by default is the name of the entity).
+     * @param {boolean} enableTranslation - If translations are enabled or not
+     * @param {string} clientFramework - The name of the client framework
+     * @param {string} entityTranslationKeyMenu - i18n key for entity entry in menu
+     */
```
Comment (MathieuAA) [inline generators/generator-base.js]: missing parameter: `entityModule`

File: generators/generator-base.js [hunk 9]
```diff
@@ -101,15 +101,33 @@ module.exports = class extends PrivateBase {
...
+    addElementToAnyMenu(routerName, enableTranslation, clientFramework, entityTranslationKeyMenu, entityModule) {
```
Comment (MathieuAA) [inline generators/generator-base.js]: Refactor hint:
    - there are 5 parameters (which may be too much)
    - everytime we use the entityModule param to determine what method to call with the other 4
    - the 4 params might belong to the same object (or structure) and passing them in that order, everytime, isn't that pretty

File: generators/client/needle-api/needle-client-angular.js [hunk 10]
```diff
@@ -182,33 +238,26 @@ module.exports = class extends needleClientBase {
...
-                    : `|{
-                                |                path: '${entityUrl}',
-                                |                loadChildren: () => import('${modulePath}').then(m => m.${moduleName})
-                                |            }`;
-                const rewriteFileModel = this.generateFileModel(
-                    entityModulePath,
-                    'jhipster-needle-add-entity-route',
-                    this.generator.stripMargin(splicable)
-                );
+                let splicable = addComma ? ',' : '';
```
Comment (MathieuAA) [inline generators/client/needle-api/needle-client-angular.js]: I don't understand what splicable means, do you mean spliceable?

**Comment by mshima** [comment]
Thank you @MathieuAA.
Closing this because it is too complicated for a simple thing.

**Pull request closed without merging**